### PR TITLE
WIP Optimization of SynchronizeLDM

### PR DESCRIPTION
**Pull request by panjan**

### CODE REVIEW

File: lib/gooddata/models/project.rb [hunk 1]
```diff
@@ -1860,6 +1871,10 @@ def update_from_blueprint(blueprint, options = {})
       GoodData::Model::ProjectCreator.migrate(options.merge(spec: blueprint, token: options[:auth_token], client: client, project: self))
     end
 
+    def update_from_blueprint_async(blueprint, options = {})
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Similar code found in project.rb:1870, project.rb:1874 (mass = 42)

File: lib/gooddata/models/project_creator.rb [hunk 2]
```diff
@@ -13,6 +13,27 @@ module GoodData
...
+          opts = { client: GoodData.connection, execute_ca_scripts: true }.merge(opts)
+          client = opts[:client]
+          fail ArgumentError, 'No :client specified' if client.nil?
+
+          spec = opts[:spec] || fail('You need to provide spec for migration')
+          bp = ProjectBlueprint.new(spec)
+          fail GoodData::ValidationError, "Blueprint is invalid #{bp.validate.inspect}" unless bp.valid?
+          spec = bp.to_hash
+
+          project = opts[:project] || client.create_project(opts.merge(:title => opts[:title] || spec[:title], :client => client, :environment => opts[:environment]))
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Identical code found in project_creator.rb:26, project_creator.rb:47 (mass = 96)

File: lib/gooddata/models/project_creator.rb [hunk 3]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+          to_poll = []
+          unless dry_run
+            errors = []
+            replaced_maqls.each do |replaced_maql_chunks|
+              begin
+                replaced_maql_chunks['updateScript']['maqlDdlChunks'].each do |chunk|
+                  GoodData.logger.debug(chunk)
+                  to_poll << project.execute_maql_async(chunk)
+                end
+              rescue => e
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Identical code found in project_creator.rb:86, project_creator.rb:139 (mass = 84)

Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Has the variable name 'e' (UncommunicativeVariableName)

File: lib/gooddata/models/project_creator.rb [hunk 4]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+                puts "Error occured when executing MAQL, project: \"#{project.title}\" reason: \"#{e.message}\", chunks: #{replaced_maql_chunks.inspect}"
+                errors << e
+                next
+              end
+            end
+
+            if ca_chunks && opts[:execute_ca_scripts]
+              begin
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Identical code found in project_creator.rb:94, project_creator.rb:147 (mass = 136)

File: lib/gooddata/models/project_creator.rb [hunk 5]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+                ca_chunks.each { |chunk| project.execute_maql(chunk) }
+              rescue => e
+                puts "Error occured when executing MAQL, project: \"#{project.title}\" reason: \"#{e.message}\", chunks: #{ca_chunks.inspect}"
+                errors << e
+              end
+            end
+
+            if (!errors.empty?) && (errors.length == replaced_maqls.length)
+              messages = errors.map { |err| GoodData::Helpers.interpolate_error_messages(err.data['wTaskStatus']['messages']) }
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Identical code found in project_creator.rb:103, project_creator.rb:156 (mass = 124)

File: lib/gooddata/lcm/actions/synchronize_ldm.rb [hunk 6]
```diff
@@ -41,32 +41,49 @@ def call(params)
...
 
-              results << {
-                from: from_project,
-                to: pid,
-                status: 'ok'
-              }
-              entry
+            to_poll.each do |polling_uri|
+              result = client.poll_on_response(polling_uri, options) do |body|
+                body && body['wTaskStatus'] && body['wTaskStatus']['status'] == 'RUNNING'
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_ldm.rb]: Calls 'body['wtaskstatus']' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_ldm.rb [hunk 7]
```diff
@@ -41,32 +41,49 @@ def call(params)
...
-
-              entry[:ca_scripts] = ca_scripts
+          synchronize = []
+          params.synchronize.each_slice(100) do |slice|
+            synchronize_result = slice.synchronize.map do |info|
+              from_project = info.from
+              to_projects = info.to
+
+              from = development_client.projects(from_project) || fail("Invalid 'from' project specified - '#{from_project}'")
+              params.gdc_logger.info "Creating Blueprint, project: '#{from.title}', PID: #{from.pid}"
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_ldm.rb]: Calls 'params.gdc_logger' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_ldm.rb [hunk 8]
```diff
@@ -41,32 +41,49 @@ def call(params)
...
+
+              blueprint = from.blueprint(include_ca: include_ca)
+              to_poll = []
+              info[:to] = to_projects.pmap do |entry|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_ldm.rb]: Contains iterators nested 3 deep (NestedIterators)

File: lib/gooddata/models/project.rb [hunk 9]
```diff
@@ -837,6 +837,17 @@ def execute_maql(maql, options = {})
       result
     end
 
+    # Executes MAQL expression
+    #
+    # @param maql [String] MAQL expression
+    # @return [String] URI to poll
+    def execute_maql_async(maql, options = {})
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Has unused parameter 'options' (UnusedParameters)

Comment (yenkins) [inline lib/gooddata/models/project.rb]: Lint/UnusedMethodArgument: Unused method argument - `options`. If it's necessary, use `_` or `_options` as an argument name to indicate that it won't be used.

File: lib/gooddata/models/project_creator.rb [hunk 10]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+          opts = { client: GoodData.connection }.merge(opts)
+          dry_run = opts[:dry_run]
+          replacements = opts['maql_replacements'] || opts[:maql_replacements] || {}
+
+          _, project = GoodData.get_client_and_project(opts)
+
+          bp = ProjectBlueprint.new(spec)
+          response = project.maql_diff(blueprint: bp, params: [:includeGrain])
+
+          GoodData.logger.debug("projectModelDiff") { response.pretty_inspect }
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Calls 'gooddata.logger' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/project_creator.rb [hunk 11]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+          unless dry_run
+            errors = []
+            replaced_maqls.each do |replaced_maql_chunks|
+              begin
+                replaced_maql_chunks['updateScript']['maqlDdlChunks'].each do |chunk|
+                  GoodData.logger.debug(chunk)
+                  to_poll << project.execute_maql_async(chunk)
+                end
+              rescue => e
+                puts "Error occured when executing MAQL, project: \"#{project.title}\" reason: \"#{e.message}\", chunks: #{replaced_maql_chunks.inspect}"
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Calls 'e.message' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Calls 'project.title' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/project_creator.rb [hunk 12]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+                errors << e
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Calls 'errors << e' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/project_creator.rb [hunk 13]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+          dry_run = opts[:dry_run]
+          replacements = opts['maql_replacements'] || opts[:maql_replacements] || {}
+
+          _, project = GoodData.get_client_and_project(opts)
+
+          bp = ProjectBlueprint.new(spec)
+          response = project.maql_diff(blueprint: bp, params: [:includeGrain])
+
+          GoodData.logger.debug("projectModelDiff") { response.pretty_inspect }
+          chunks = response['projectModelDiff']['updateScripts']
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Calls 'response['projectmodeldiff']' 3 times (DuplicateMethodCall)

File: lib/gooddata/models/project_creator.rb [hunk 14]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+          return [] if chunks.empty?
+
+          ca_maql = response['projectModelDiff']['computedAttributesScript'] if response['projectModelDiff']['computedAttributesScript']
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Calls 'response['projectmodeldiff']['computedattributesscript']' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/project_creator.rb [hunk 15]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
...
+
+          maqls = pick_correct_chunks(chunks, opts)
+          replaced_maqls = apply_replacements_on_maql(maqls, replacements)
+
+          to_poll = []
+          unless dry_run
+            errors = []
+            replaced_maqls.each do |replaced_maql_chunks|
+              begin
+                replaced_maql_chunks['updateScript']['maqlDdlChunks'].each do |chunk|
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Contains iterators nested 2 deep (NestedIterators)

File: lib/gooddata/models/project_creator.rb [hunk 16]
```diff
@@ -13,6 +13,27 @@ module GoodData
   module Model
     class ProjectCreator
       class << self
+        def migrate_async(opts = {})
+          opts = { client: GoodData.connection, execute_ca_scripts: true }.merge(opts)
+          client = opts[:client]
+          fail ArgumentError, 'No :client specified' if client.nil?
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/models/project_creator.rb [hunk 17]
```diff
@@ -13,6 +13,27 @@ module GoodData
   module Model
     class ProjectCreator
       class << self
+        def migrate_async(opts = {})
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Has approx 15 statements (TooManyStatements)

File: lib/gooddata/models/project_creator.rb [hunk 18]
```diff
@@ -86,6 +107,59 @@ def migrate_datasets(spec, opts = {})
           replaced_maqls + (ca_maql ? [ca_maql] : [])
         end
 
+        def migrate_datasets_async(spec, opts = {})
```
Comment (yenkins) [inline lib/gooddata/models/project_creator.rb]: Has approx 33 statements (TooManyStatements)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/638/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/430/ : FAILURE in 2m 07s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/738/ : FAILURE in 1m 51s

**Comment by panjan** [comment]
Running maql executions in parallel doesn't have a big effect as they are queued on one dwh worker anyway.

**Pull request closed without merging**